provider/provider_test.go

```diff
@@ -587,3 +675,290 @@ func TestCanPreview(t *testing.T) {
+					}),
+				}),
+			},
+			news: resource.PropertyMap{
+				"registryAuth": resource.NewArrayProperty([]resource.PropertyValue{
+					resource.NewObjectProperty(resource.PropertyMap{
+						"address":  resource.NewStringProperty("dockerhub"),
+						"username": resource.NewStringProperty("new-username"),
+						"password": resource.NewStringProperty("new-password"),
+					}),
+					resource.NewObjectProperty(resource.PropertyMap{
+						"address":  resource.NewStringProperty("ecr"),
+						"username": resource.NewStringProperty("user"),
+						"password": resource.NewStringProperty("pass"),
+					}),
+				}),
+			},
+			want: &rpc.DiffResponse{
+				Changes: rpc.DiffResponse_DIFF_NONE,
+			},
+		},
+		{
+			name: "Changes for computed in registryAuth",
+			olds: resource.PropertyMap{
+				"registryAuth": resource.NewArrayProperty([]resource.PropertyValue{
+					resource.NewObjectProperty(resource.PropertyMap{
+						"address":  resource.NewStringProperty("dockerhub"),
+						"username": resource.NewStringProperty("username"),
+						"password": resource.NewStringProperty("password"),
+					}),
+					resource.NewObjectProperty(resource.PropertyMap{
+						"address":  resource.NewStringProperty("ecr"),
+						"username": resource.NewStringProperty("user"),
+						"password": resource.NewStringProperty("pass"),
+					}),
+				}),
+			},
+			news: resource.PropertyMap{
+				"registryAuth": resource.NewComputedProperty(resource.Computed{
+					Element: resource.NewArrayProperty([]resource.PropertyValue{
+						resource.NewObjectProperty(resource.PropertyMap{
+							"address":  resource.NewStringProperty("dockerhub"),
+							"username": resource.NewStringProperty("username"),
+							"password": resource.NewStringProperty("password"),
+						}),
+						resource.NewObjectProperty(resource.PropertyMap{
+							"address":  resource.NewStringProperty("ecr"),
+							"username": resource.NewStringProperty("user"),
+							"password": resource.NewStringProperty("pass"),
+						}),
+					}),
+				}),
+			},
+			want: &rpc.DiffResponse{
+				Changes: rpc.DiffResponse_DIFF_SOME,
+				DetailedDiff: map[string]*rpc.PropertyDiff{
+					"registryAuth": {
+						Kind: rpc.PropertyDiff_UPDATE,
+					},
+				},
+				HasDetailedDiff: true,
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			p := &dockerNativeProvider{}
+
+			olds, err := plugin.MarshalProperties(tt.olds, plugin.MarshalOptions{KeepUnknowns: true})
+			require.NoError(t, err)
+
+			news, err := plugin.MarshalProperties(tt.news, plugin.MarshalOptions{KeepUnknowns: true})
+			require.NoError(t, err)
+
+			req := &rpc.DiffRequest{
+				Urn:  string("urn:pulumi:test::docker-provider::docker:index/image:Image::foo"),
+				Olds: olds,
+				News: news,
+			}
+
+			got, err := p.Diff(context.TODO(), req)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("dockerNativeProvider.Diff() error = %v, wantErr %v", err, tt.wantErr)
```

Comment:
This is a very confusing test suite we are testing dockerNativeProvider.Diff; the only job of dockerNativeProvider.Diff is to compute diffs for the Image resource. However we're testing it with registryAuth field that is not even part of the Image resource schema. I think we can keep the test suite but target 

```
func (dp dockerHybridProvider) DiffConfig(ctx context.Context, request *rpc.DiffRequest) (*rpc.DiffResponse, error) {
```

Reply:
Yeah, more reason to split out the methods!